src/main/java/org/tron/p2p/connection/socket/PeerClient.java

```diff
@@ -41,38 +41,58 @@ public void close() {
   public void connect(String host, int port, String remoteId) {
     try {
       ChannelFuture f = connectAsync(host, port, remoteId, false, false);
-      f.sync().channel().closeFuture().sync();
+      if (f != null) {
+        f.sync().channel().closeFuture().sync();
+      }
     } catch (Exception e) {
       log.warn("PeerClient can't connect to {}:{} ({})", host, port, e.getMessage());
     }
   }
 
   public ChannelFuture connect(Node node, ChannelFutureListener future) {
-    ChannelFuture channelFuture = connectAsync(
-        node.getPreferInetSocketAddress().getAddress().getHostAddress(),
-        node.getPort(),
-        node.getId() == null ? Hex.toHexString(NetUtil.getNodeId()) : node.getHexId(), false, false);
-    if (future != null) {
-      channelFuture.addListener(future);
+    ChannelFuture channelFuture;
+    try {
+      channelFuture = connectAsync(
+          node.getPreferInetSocketAddress().getAddress().getHostAddress(),
+          node.getPort(),
+          node.getId() == null ? Hex.toHexString(NetUtil.getNodeId()) : node.getHexId(), false,
+          false);
+      if (channelFuture != null && future != null) {
+        channelFuture.addListener(future);
+      }
+    } catch (Exception e) {
+      log.warn("PeerClient can't connect to {} ({})", node.getPreferInetSocketAddress(),
+          e.getMessage());
+      return null;
     }
     return channelFuture;
   }
 
   public ChannelFuture connectAsync(Node node, boolean discoveryMode) {
-    return connectAsync(node.getPreferInetSocketAddress().getAddress().getHostAddress(),
-        node.getPort(),
-        node.getId() == null ? Hex.toHexString(NetUtil.getNodeId()) : node.getHexId(),
-        discoveryMode, true)
-        .addListener((ChannelFutureListener) future -> {
-          if (!future.isSuccess()) {
-            log.warn("Connect to peer {} fail, cause:{}", node.getPreferInetSocketAddress(),
-                future.cause().getMessage());
-            future.channel().close();
-            if (!discoveryMode) {
-              ChannelManager.triggerConnect(node.getPreferInetSocketAddress());
-            }
+    ChannelFuture channelFuture;
```

Comment:
What to do if the channelFuture is created successfully but no listener is added？

Reply:
This will happen if and only if the process is shut down.  Any following opertion is meaningless.